src/client.rs

```diff
@@ -55,27 +60,42 @@ pub fn pair<Req, Resp, ReqBody, E>() -> Pair<Req, Resp, ReqBody, E>
     (client, rx)
 }
 
-impl<Req, Resp, ReqBody, E> Service for Client<Req, Resp, ReqBody, E>
-    where Req: Send + 'static,
-          Resp: Send + 'static,
-          ReqBody: Stream<Error = E>,
-          E: From<Error<E>> + Send + 'static,
+impl<R1, R2, B1, B2, E> Service for Client<R1, R2, B1, B2, E>
+    where R1: 'static,
+          R2: 'static,
+          B1: Stream<Error = E> + 'static,
+          B2: 'static,
+          E: From<Error<E>> + 'static,
 {
-    type Request = Message<Req, ReqBody>;
-    type Response = Resp;
+    type Request = Message<R1, B1>;
+    type Response = Message<R2, B2>;
     type Error = E;
-    type Future = BoxFuture<Self::Response, E>;
+    type Future = Response<Self::Response, E>;
 
     fn call(&self, request: Self::Request) -> Self::Future {
         let (tx, rx) = futures::oneshot();
 
         // TODO: handle error
         self.tx.borrow_mut().send(Ok((request, tx)));
 
-        rx.then(|t| t.unwrap()).boxed()
+        Response { inner: rx }
     }
 
     fn poll_ready(&self) -> Async<()> {
         self.tx.borrow_mut().poll_ready().unwrap()
     }
 }
+
+impl<T, E> Future for Response<T, E> {
+    type Item = T;
+    type Error = E;
+
+    fn poll(&mut self) -> Poll<T, E> {
+        match self.inner.poll() {
+            Ok(Async::Ready(Ok(v))) => Ok(Async::Ready(v)),
+            Ok(Async::Ready(Err(e))) => Err(e),
+            Ok(Async::NotReady) => Ok(Async::NotReady),
+            Err(_) => panic!("aborted"),
```

Comment:
Not sure. I added the panic to move on and come back to later. I **think** that right now this panic shouldn't be reachable. However, at some point when the reactor shutdown story is figured out, it seems plausible that the case will be real. If the reactor is shutdown and resources freed, the pipeline/multiplex task will be dropped, which will cause all streams / oneshots to be canceled.

Anyway, I'm not sure what the best thing to do right now is.